lib/plug/csrf_protection.ex

```diff
@@ -103,17 +103,34 @@ defmodule Plug.CSRFProtection do
 
   def init(opts), do: opts
 
-  def call(conn, _opts) do
+  def call(conn, opts) do
+    opts = verify_opts opts
     csrf_token = get_session(conn, "_csrf_token")
     Process.put(:plug_csrf_token, csrf_token)
-
     if not verified_request?(conn, csrf_token) do
-      raise InvalidCSRFTokenError
+      if opts[:with] == :exception do
```

Comment:
Since `opts` value is validated in `init/1` function there is might be a simple `case` call:
```elixir
case Keyword.get(opts, :with) do
  :exception   -> raise InvalidCSRFTokenError
  :nil_session -> clear_session(conn)
end
```